### ISSUE TITLE
Has GCCs -fPIC behavior changed?

### ISSUE THREAD

**Issue by Paebbels** [tgingold/ghdl#213]
Hello,

I did a fresh GHDL build and executed a simple testbench. I get the following output. Please see the second half (elaboration), where `/usr/bin/ld` complains about `-fPIC`.

```Bash
paebbels@debian8:~/git/PoC$ ./poc.sh -v ghdl arith.prng
================================================================================
                         The PoC-Library - Service Tool                         
================================================================================
Initializing PoC-Library Service Tool for simulations
  Reading configuration files...
Preparing simulation environment...
  Creating fresh temporary directory.
  Changing working directory to temporary directory.
  Preparing GHDL simulator.
Testbench: PoC.arith.prng
  Creating PoC project 'arith_prng_tb'
  Reading filelist '/home/paebbels/git/PoC/tb/arith/arith_prng_tb.files'
Running analysis for every vhdl file...
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/tb/common/my_project.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/tb/common/my_config_GENERIC.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/common/utils.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/common/config.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/common/math.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/common/strings.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/common/vectors.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/common/physical.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/common/components.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/common/protected.v08.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/common/fileio.v08.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/sim/sim_types.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/sim/sim_random.v08.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/sim/sim_protected.v08.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/sim/sim_global.v08.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/sim/sim_simulation.v08.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/sim/sim_waveform.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/arith/arith.pkg.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=poc /home/paebbels/git/PoC/src/arith/arith_prng.vhdl
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl --ieee=synopsys --std=08 --work=test /home/paebbels/git/PoC/tb/arith/arith_prng_tb.vhdl
Running elaboration...
  command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -e -fexplicit --ieee=synopsys --std=08 --work=test arith_prng_tb
  ghdl elaborate messages for 'test.arith_prng_tb'
  ------------------------------------------------------------------------------
  /usr/bin/ld: e~arith_prng_tb.o: relocation R_X86_64_32 against `.bss' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: /opt/ghdl/0.34-dev-llvm/lib/ghdl/v08/std/std_standard.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: /opt/ghdl/0.34-dev-llvm/lib/ghdl/v08/std/textio.o: relocation R_X86_64_32 against symbol `std__textio__RTI' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: /opt/ghdl/0.34-dev-llvm/lib/ghdl/v08/std/textio_body.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: /opt/ghdl/0.34-dev-llvm/lib/ghdl/v08/synopsys/std_logic_1164.o: relocation R_X86_64_32 against symbol `ieee__std_logic_1164__RTI' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: /opt/ghdl/0.34-dev-llvm/lib/ghdl/v08/synopsys/std_logic_1164-body.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: /opt/ghdl/0.34-dev-llvm/lib/ghdl/v08/synopsys/numeric_std.o: relocation R_X86_64_32 against symbol `ieee__numeric_std__RTI' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: /opt/ghdl/0.34-dev-llvm/lib/ghdl/v08/synopsys/numeric_std-body.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: /opt/ghdl/0.34-dev-llvm/lib/ghdl/v08/synopsys/math_real.o: relocation R_X86_64_32 against symbol `ieee__math_real__RTI' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: /opt/ghdl/0.34-dev-llvm/lib/ghdl/v08/synopsys/math_real-body.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: utils.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: my_config_GENERIC.o: relocation R_X86_64_32 against symbol `poc__my_config__RTI' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: my_project.o: relocation R_X86_64_32 against symbol `poc__my_project__RTI' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: config.o: relocation R_X86_64_32S against symbol `poc__config_private__t_board_config_string__STB' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: strings.o: relocation R_X86_64_32S against symbol `poc__strings__t_rawchar__STB' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: vectors.o: relocation R_X86_64_32S against symbol `poc__vectors__t_slv_2__STB' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: physical.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: sim_types.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: protected.v08.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: fileio.v08.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: sim_protected.v08.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: sim_global.v08.o: relocation R_X86_64_32 against symbol `poc__sim_global__RTI' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: sim_simulation.v08.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: sim_waveform.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: arith_prng_tb.o: relocation R_X86_64_32S against symbol `test__arith_prng_tb__ARCH__tb__RTI' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: arith_prng.o: relocation R_X86_64_32 against symbol `poc__arith_prng__clk__RTI' can not be used when making a shared object; recompile with -fPIC
  /usr/bin/ld: final link failed: Nonrepresentable section on output
  collect2: error: ld returned 1 exit status
  /opt/ghdl/0.34-dev-llvm/bin/ghdl: compilation error
  ------------------------------------------------------------------------------
  ERROR: Error while elaborating 'test.arith_prng_tb'.
  [SKIPPED DUE TO ERRORS]
================================================================================
                           Overall Simulation Report                            
================================================================================
Name                     |  Time |   Status   
--------------------------------------------------------------------------------
arith                    |       | 
  prng                   |  0:02 | ELAB. ERROR
================================================================================
Time:  0:02  Count: 1    Passed: 0    No Asserts: 0   Failed: 0   Errors: 1 
================================================================================
```

I know GCC has two PIC options: `-fPIC` and `-fpic`. Has something chainged in GHDL's makefiles, GCC's default PIC behavior or in GNAT?

**GHDL version:**
```Bash
paebbels@debian8:~/git/PoC$ /opt/ghdl/0.34-dev-llvm/bin/ghdl -v
GHDL 0.34-dev (2016-09-14-222-g7db0826-dirty) [Dunoon edition]
 Compiled with GNAT Version: 6.2.0
 llvm code generator
```

**Addendum 1:**
Setting `CC=gcc-4.9` while configuring doesn't make any difference, because most instructions generated by make are using the system default GCC, which is 6.2.0. Only the C code is compiled with my 4.9.3.

```make
...
gcc-6 -c -I./ -I../../src/ortho/llvm -I../../src/ortho -I../../src -I../../src/vhdl -I../../src/psl -I../../src/vhdl/translate -I../../src/ghdldrv -I../../src/grt -I../../src/ortho -I../../src/ortho/llvm -gnaty3befhkmr -gnatwae -gnatf -gnat05 -g -gnata -I- /home/paebbels/git/ghdl/src/vhdl/iirs_walk.adb
gnatbind-6 -aI../../src/ortho/llvm -aI../../src/ortho -aI../../src -aI../../src/vhdl -aI../../src/psl -aI../../src/vhdl/translate -aI../../src/ghdldrv -aI../../src/grt -aI../../src/ortho -aI../../src/ortho/llvm -aO. -E -x ortho_code_main.ali
gnatlink-6 ortho_code_main.ali -o ghdl1-llvm -g llvm-cbindings.o --LINK=clang++ -L/usr/lib/llvm-3.5/lib -lLLVMLTO -lLLVMObjCARCOpts -lLLVMLinker -lLLVMipo -lLLVMVectorize -lLLVMBitWriter -lLLVMIRReader -lLLVMAsmParser -lLLVMR600CodeGen -lLLVMR600Desc -lLLVMR600Info -lLLVMR600AsmPrinter -lLLVMSystemZDisassembler -lLLVMSystemZCodeGen -lLLVMSystemZAsmParser -lLLVMSystemZDesc -lLLVMSystemZInfo -lLLVMSystemZAsmPrinter -lLLVMHexagonCodeGen -lLLVMHexagonAsmPrinter -lLLVMHexagonDesc -lLLVMHexagonInfo -lLLVMNVPTXCodeGen -lLLVMNVPTXDesc -lLLVMNVPTXInfo -lLLVMNVPTXAsmPrinter -lLLVMCppBackendCodeGen -lLLVMCppBackendInfo -lLLVMMSP430CodeGen -lLLVMMSP430Desc -lLLVMMSP430Info -lLLVMMSP430AsmPrinter -lLLVMXCoreDisassembler -lLLVMXCoreCodeGen -lLLVMXCoreDesc -lLLVMXCoreInfo -lLLVMXCoreAsmPrinter -lLLVMMipsDisassembler -lLLVMMipsCodeGen -lLLVMMipsAsmParser -lLLVMMipsDesc -lLLVMMipsInfo -lLLVMMipsAsmPrinter -lLLVMAArch64Disassembler -lLLVMAArch64CodeGen -lLLVMAArch64AsmParser -lLLVMAArch64Desc -lLLVMAArch64Info -lLLVMAArch64AsmPrinter -lLLVMAArch64Utils -lLLVMARMDisassembler -lLLVMARMCodeGen -lLLVMARMAsmParser -lLLVMARMDesc -lLLVMARMInfo -lLLVMARMAsmPrinter -lLLVMPowerPCDisassembler -lLLVMPowerPCCodeGen -lLLVMPowerPCAsmParser -lLLVMPowerPCDesc -lLLVMPowerPCInfo -lLLVMPowerPCAsmPrinter -lLLVMSparcDisassembler -lLLVMSparcCodeGen -lLLVMSparcAsmParser -lLLVMSparcDesc -lLLVMSparcInfo -lLLVMSparcAsmPrinter -lLLVMTableGen -lLLVMDebugInfo -lLLVMOption -lLLVMX86Disassembler -lLLVMX86AsmParser -lLLVMX86CodeGen -lLLVMSelectionDAG -lLLVMAsmPrinter -lLLVMX86Desc -lLLVMX86Info -lLLVMX86AsmPrinter -lLLVMX86Utils -lLLVMJIT -lLLVMLineEditor -lLLVMMCAnalysis -lLLVMMCDisassembler -lLLVMInstrumentation -lLLVMInterpreter -lLLVMCodeGen -lLLVMScalarOpts -lLLVMInstCombine -lLLVMTransformUtils -lLLVMipa -lLLVMAnalysis -lLLVMProfileData -lLLVMMCJIT -lLLVMTarget -lLLVMRuntimeDyld -lLLVMObject -lLLVMMCParser -lLLVMBitReader -lLLVMExecutionEngine -lLLVMMC -lLLVMCore -lLLVMSupport -lz -lpthread -lffi -ledit -ltinfo -ldl -lm
clang: warning: argument unused during compilation: '-shared-libgcc'
make[1]: Verzeichnis „/home/paebbels/git/ghdl/build/llvm“ wird verlassen
gnatmake -aI../../src -aI../../src/vhdl -aI../../src/psl -aI../../src/vhdl/translate -aI../../src/ghdldrv -aI../../src/grt -aI../../src/ortho -aI../../src/ortho/llvm -aI../../src/ghdldrv -gnaty3befhkmr -gnatwae -aO. -gnatf -gnat05 -g -gnata\
 ghdl_llvm -bargs -E -largs 
gcc-6 -c -I./ -I../../src -I../../src/vhdl -I../../src/psl -I../../src/vhdl/translate -I../../src/ghdldrv -I../../src/grt -I../../src/ortho -I../../src/ortho/llvm -I../../src/ghdldrv -gnaty3befhkmr -gnatwae -gnatf -gnat05 -g -gnata -I- /home/paebbels/git/ghdl/src/ghdldrv/ghdl_llvm.adb
gcc-6 -c -I./ -I../../src -I../../src/vhdl -I../../src/psl -I../../src/vhdl/translate -I../../src/ghdldrv -I../../src/grt -I../../src/ortho -I../../src/ortho/llvm -I../../src/ghdldrv -gnaty3befhkmr -gnatwae -gnatf -gnat05 -g -gnata -I- /home/paebbels/git/ghdl/src/ghdldrv/ghdldrv.adb
gcc-6 -c -I./ -I../../src -I../../src/vhdl -I../../src/psl -I../../src/vhdl/translate -I../../src/ghdldrv -I../../src/grt -I../../src/ortho -I../../src/ortho/llvm -I../../src/ghdldrv -gnaty3befhkmr -gnatwae -gnatf -gnat05 -g -gnata -I- /home/paebbels/git/ghdl/src/ghdldrv/ghdllocal.adb
gcc-6 -c -I./ -I../../src -I../../src/vhdl -I../../src/psl -I../../src/vhdl/translate -I../../src/ghdldrv -I../../src/grt -I../../src/ortho -I../../src/ortho/llvm -I../../src/ghdldrv -gnaty3befhkmr -gnatwae -gnatf -gnat05 -g -gnata -I- /home/paebbels/git/ghdl/src/ghdldrv/ghdlmain.adb
gcc-6 -c -I./ -I../../src -I../../src/vhdl -I../../src/psl -I../../src/vhdl/translate -I../../src/ghdldrv -I../../src/grt -I../../src/ortho -I../../src/ortho/llvm -I../../src/ghdldrv -gnaty3befhkmr -gnatwae -gnatf -gnat05 -g -gnata -I- /home/paebbels/git/ghdl/src/ghdldrv/ghdlprint.adb
gcc-6 -c -I./ -I../../src -I../../src/vhdl -I../../src/psl -I../../src/vhdl/translate -I../../src/ghdldrv -I../../src/grt -I../../src/ortho -I../../src/ortho/llvm -I../../src/ghdldrv -gnaty3befhkmr -gnatwae -gnatf -gnat05 -g -gnata -I- /home/paebbels/git/ghdl/src/ghdldrv/ghdlvpi.adb
gcc-6 -c -I./ -I../../src -I../../src/vhdl -I../../src/psl -I../../src/vhdl/translate -I../../src/ghdldrv -I../../src/grt -I../../src/ortho -I../../src/ortho/llvm -I../../src/ghdldrv -gnaty3befhkmr -gnatwae -gnatf -gnat05 -g -gnata -I- /home/paebbels/git/ghdl/src/ghdldrv/ghdlxml.adb
gcc-6 -c -I../../src -I../../src/vhdl -I../../src/psl -I../../src/vhdl/translate -I../../src/ghdldrv -I../../src/grt -I../../src/ortho -I../../src/ortho/llvm -I../../src/ghdldrv -gnaty3befhkmr -gnatwae -gnatf -gnat05 -g -gnata default_pathes.ads
gnatbind-6 -aI../../src -aI../../src/vhdl -aI../../src/psl -aI../../src/vhdl/translate -aI../../src/ghdldrv -aI../../src/grt -aI../../src/ortho -aI../../src/ortho/llvm -aI../../src/ghdldrv -aO. -E -x ghdl_llvm.ali
gnatlink-6 ghdl_llvm.ali -g
gcc-4.9 -c -g  -o jumps.o ../../src/grt/config/jumps.c
gcc-4.9 -c -g -o times.o ../../src/grt/config/times.c
gcc-4.9 -c -g -o grt-cbinding.o ../../src/grt/grt-cbinding.c
gcc-4.9 -c -g -o grt-cvpi.o ../../src/grt/grt-cvpi.c
gcc-4.9 -c -g -o fstapi.o -I../../src/grt/fst ../../src/grt/fst/fstapi.c
gcc-4.9 -c -g -o lz4.o ../../src/grt/fst/lz4.c
gcc-4.9 -c -g -o fastlz.o ../../src/grt/fst/fastlz.c
gnatmake -c -aI../../src/grt -gnatec../../src/grt/grt.adc -gnat05 -gnatdY \
  ghdl_main  -cargs -g
gcc-6 -c -I./ -I../../src/grt -gnatec../../src/grt/grt.adc -gnat05 -gnatdY -g -I- /home/paebbels/git/ghdl/src/grt/ghdl_main.adb
gcc-6 -c -I./ -I../../src/grt -gnatec../../src/grt/grt.adc -gnat05 -gnatdY -g -I- /home/paebbels/git/ghdl/src/grt/grt.ads
gcc-6 -c -I./ -I../../src/grt -gnatec../../src/grt/grt.adc -gnat05 -gnatdY -g -I- /home/paebbels/git/ghdl/src/grt/grt-errors.adb
gcc-6 -c -I./ -I../../src/grt -gnatec../../src/grt/grt.adc -gnat05 -gnatdY -g -I- /home/paebbels/git/ghdl/src/grt/grt-main.adb
...
```

**Addendum 2:**
> The new toolchains of Ubuntu (and Debian soon too) default to PIE code, so in order to link correctly, the project needs to be compiled with PIE/PIC to work.

Posted on: October 13, 2016  
Source: http://forum.dlang.org/thread/tppsgztsbsdrtkpcbtak@forum.dlang.org?page=1

**Addendum 3:**

I implemented 3 new GHDL options in PoC so I can pass `-Wc`, `-Wa` and `-Wl` to GHDL. When doing so, I get these results, saying that `-fPIC` is unknown:

```Bash
 command: /opt/ghdl/0.34-dev-llvm/bin/ghdl -a -v -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl -Wc,-fPIC --ieee=synopsys --std=08 --work=test /home/paebbels/git/PoC/tb/arith/arith_prng_tb.vhdl
  ghdl analyze messages for '/home/paebbels/git/PoC/tb/arith/arith_prng_tb.vhdl'
  ------------------------------------------------------------------------------
  /opt/ghdl/0.34-dev-llvm/bin/ghdl1-llvm -fexplicit -frelaxed-rules --warn-binding --no-vital-checks --mb-comments --syn-binding -fpsl -fPIC --std=08 --work=test -P/opt/ghdl/0.34-dev-llvm/lib/ghdl/v08/std/ -P/opt/ghdl/0.34-dev-llvm/lib/ghdl/v08/synopsys/ -c -o arith_prng_tb.o /home/paebbels/git/PoC/tb/arith/arith_prng_tb.vhdl
  unknown option '-fPIC'
```

* Line 1 is printed by PoC
* Line 4 is printed by `bin/ghdl`
* Line 5 is caused by `bin/ghdl1-llvm`

`-fPIC` is a compiler option, right?

**Comment by tgingold** [comment]
On 02/12/16 20:01, Patrick Lehmann wrote:
> Hello,
>
> I did ne fresh GHDL build and executed a simple testbench. I get the
> following output. Please see the second half (elaboration), where
> |/usr/bin/ld| complains about |-fPIC|.

Did you upgrade your system recently ?

Try to link with -no-pie

Regard,
Tristan.

**Comment by Paebbels** [comment]
Hmmm, my test system is a Debian testing, it automatically uses new software releases. Unfortunately, they declared GCC 6.2.0-13 as the new default and jumped from 5.4 over 6.0 (and 6.1?) directly to 6.2 with a zero in the release number ...

**So I'll use this GHDL elaborate call:**
```Bash
/opt/ghdl/0.34-dev-llvm/bin/ghdl -e -v -fexplicit -Wl,-no-pie --ieee=synopsys --std=08 --work=test arith_prng_tb
```

OK this works.

**Some questions remain:**
* Is this a workaround or solution?
* What back ends are effected?
* Should GHDL compile with `-fPIC` by default?

**Comment by tgingold** [comment]
On 03/12/16 17:27, Patrick Lehmann wrote:
> Hmmm, my test system is a Debian testing, it automatically uses new
> software releases. Unfortunately, they declared GCC 6.2.0-13 as the new
> default and jumped from 5.4 over 6.0 (and 6.1?) directly to 6.2 with a
> zero in the release number ...
>
> *So I'll use this GHDL elaborate call:*
>
> /opt/ghdl/0.34-dev-llvm/bin/ghdl -e -v -fexplicit -Wl,-no-pie --ieee=synopsys --std=08 --work=test arith_prng_tb
>
> OK this works.
>
> *Some questions remain:*
>
>   * Is this a workaround or solution?

Only a work-around.

>   * What back ends are effected?

I suppose that both llvm and gcc back ends are affected.

>   * Should GHDL compile with |-fPIC| by default?

Yes.  What is the output of 'gcc -v' on your machine ?

**Comment by Paebbels** [comment]
The GCC version output is:

```
paebbels@debian8:~/git/PoC$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/6/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 6.2.0-13' --with-bugurl=file:///usr/share/doc/gcc-6/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-6 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --enable-default-pie --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-6-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-6-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-6-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 6.2.0 20161109 (Debian 6.2.0-13) 
```

**Comment by tgingold** [comment]
On 04/12/16 21:34, Patrick Lehmann wrote:
> The GCC version output is:

So, there is --enable-default-pie
And you are using the llvm backend for ghdl.

**Comment by Paebbels** [comment]
Yes the "productive" GHDL version is using LLVM. Sometime I also run tests with the mcode backend. I don't use GCC any more. It's to much work to compile it and GCC itself :)

**Comment by tgingold** [comment]
On 05/12/16 02:13, Patrick Lehmann wrote:
> Yes the "productive" GHDL version is using LLVM. Sometime I also run
> tests with the mcode backend. I don't use GCC any more. It's to much
> work to compile it and GCC itself :)

So I think you have to change one line in
  src/ortho/llvm/ortho_code_main.adb:

  Reloc : constant RelocMode := RelocDefault;

-->

  Reloc : constant RelocMode := RelocPic;

Tell me if that works.

**Comment by Paebbels** [comment]
I'm looking forward to test it.

**Comment by Paebbels** [comment]
The `-no-pie` flag breaks our Travis runs ...

The GCC on Travis-CI is of version 4.8, which doesn't know anything about PIE. My problem is that I could enabled workarounds based on the GHDL version (which is always 0.34-dev), but not on the GCC version with which GHDL was build.

Just in case you need the current `gcc -v` results:
```
Testing GCC version and configuration...
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/4.8/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 4.8.4-2ubuntu1~14.04.3' --with-bugurl=file:///usr/share/doc/gcc-4.8/README.Bugs --enable-languages=c,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.8 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.8 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --disable-libmudflap --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-4.8-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-4.8-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-4.8-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 4.8.4 (Ubuntu 4.8.4-2ubuntu1~14.04.3) 


Testing GHDL version...
GHDL 0.34dev (20151126) [Dunoon edition]
 Compiled with GNAT Version: 4.8
 llvm code generator
Written by Tristan Gingold.
Copyright (C) 2003 - 2015 Tristan Gingold.
GHDL is free software, covered by the GNU General Public License.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
GHDL test [SUCCESSFUL]
```
Travis-CI report: https://travis-ci.org/VLSI-EDA/PoC/jobs/181299337#L339

**Comment by mzabeltud** [comment]
I have tested `RelocPIC` with GCC 4.9 on Debian which can also generate Position Independent Exceutables (PIE), but it is no the default there.

At first, I changed in file src/ortho/llvm/ortho_code_main.adb the line 73:

```
 Reloc : constant RelocMode := RelocDefault;
```

to

```
 Reloc : constant RelocMode := RelocPIC;
```

as suggested by @tgingold.

Now, the object files generated by LLVM have the neccessary relocation information.

But, the object files of libgrt.a must be position independ too. Thus, one has to change in the main Makefile.in, the line 72:

```
GRT_FLAGS=$(OPT_FLAGS)
```

to

```
GRT_FLAGS=$(OPT_FLAGS) -fPIC
```

Now, the LLVM GHDL simulator binary can be linked during elaboration with option `-Wl,-pie` and also without this option.

----

@Paebbels tested this with GCC 6.2 on Debian. There the linker option '-pie' is set by default.
